### ISSUE TITLE
WebsiteAgent concat with xpath not working

### ISSUE THREAD

**Issue by MakiG0** [huginn/huginn#2797]
i found huginn on reddit and i like the idea of this great tool.
At once, i had an idea: 
instead of manually reading offers, i want huginn to do that for me and compare them with an existing list of my required articles. If an item from the page matches the item from the list, then send me a message in some way. 
But at the moment I struggled with the first steps:
with the tool https://selectorgadget.com/ I have already determined the xpath of the title, subtitle and price. Huginn now gives me everything one by one instead of a single value.
This is my WebsiteAgent:

```
 {
  "expected_update_period_in_days": "2",
  "url": "https://www.kaufland.de/angebote/aktuelle-woche.category=135_Foodkn%C3%BCller.html",
  "type": "html",
  "mode": "on_change",
  "extract": {
    "title": {
      "xpath": "//*[contains(concat( \" \", @class, \" \" ), concat( \" \", \"a-pricetag__price\", \" \" ))] | //*[contains(concat( \" \", @class, \" \" ), concat( \" \", \"m-offer-tile__title\", \" \" ))] | //*[contains(concat( \" \", @class, \" \" ), concat( \" \", \"m-offer-tile__subtitle\", \" \" ))]",
      "value": "normalize-space(.)"
    }
  }
}
```

The result:
![grafik](https://user-images.githubusercontent.com/49974203/82497812-32e5be80-9aef-11ea-955c-792d5aec47f8.png)

What do I have to do to get the following result:
"Subtitle Title, price"? ( example: MAGGI Fix oder Würzpaste, 0.44)

I do not know if xpath or css is the solution.

thanks a lot in advance ;-)

**Comment by dsander** [comment]
Hi!

That is one heck of an xpath expression 😄 I think what you want is to have multiple entries in the `extract` section. One for the `title`, one for the `price`, etc.

**Comment by MakiG0** [comment]
Sorry for noob questions ;-)
I tested your solution, but it didn't work.
```
{
  "expected_update_period_in_days": "2",
  "url": "https://www.kaufland.de/angebote/aktuelle-woche.category=135_Foodkn%C3%BCller.html",
  "type": "html",
  "mode": "on_change",
  "extract": {
    "subtitle": {
      "xpath": "//*[contains(concat( \" \", @class, \" \" ), concat( \" \", \"m-offer-tile__subtitle\", \" \" ))]",
      "value": "normalize-space(.)"
    },
    "title": {
      "xpath": "//*[contains(concat( \" \", @class, \" \" ), concat( \" \", \"m-offer-tile__title\", \" \" ))]",
      "value": "normalize-space(.)"
    },
    "price": {
      "xpath": "//*[contains(concat( \" \", @class, \" \" ), concat( \" \", \"a-pricetag__price\", \" \" ))]",
      "value": "normalize-space(.)"
    }
  }
}
```
I got this error:
![grafik](https://user-images.githubusercontent.com/49974203/82711191-f2b84480-9c84-11ea-8faa-2d939f6aee7d.png)
I think, the problem is the message marked in yellow. Not every title has a subtitle and thats the problem, isn't it? 
do you have an idea?

**Comment by dsander** [comment]
Right, that is a limitation of the `WebsiteAgent`. You should be able to get around it by using two Agents. Have to first one select the wrapper of the offer like `.o-overview-list__list-item` and extract the whole HTML by specifying `.`. The second `WebsiteAgent` can process the HTML by using `data_from_event`.
